private_sharing/forms.py

```diff
@@ -26,18 +29,31 @@ class Meta:  # noqa: D101
                   'is_academic_or_nonprofit', 'add_data', 'explore_share',
                   'contact_email', 'info_url', 'short_description',
                   'long_description', 'returned_data_description', 'active',
-                  'badge_image', 'request_sources_access',
-                  'request_username_access', 'erasure_supported',
+                  'badge_image', 'request_username_access', 'erasure_supported',
                   'deauth_email_notification')
 
     def __init__(self, *args, **kwargs):
-        super(DataRequestProjectForm, self).__init__(*args, **kwargs)
+        super().__init__(*args, **kwargs)
+
+        source_projects = (DataRequestProject.objects
+                           .filter(approved=True)
+                           .exclude(returned_data_description=''))
+        if hasattr(self, 'instance'):
+            source_projects = source_projects.exclude(pk=self.instance.pk)
+            initial_grants = [rs.requested_project.id_label for rs in
+                              (RequestSourcesAccess.objects.filter(active=True)
+                              .filter(requesting_project=self.instance))]
+        else:
+            initial_grants = []
 
-        sources = [s for s in get_source_labels_and_names_including_dynamic()
-                   if s[0] != self.instance.id_label]
+        sources = [(project.id_label, project.name)
```

Comment:
is there a more straightforward way to do this stuff, if you convert the related object's field to a many-to-many?